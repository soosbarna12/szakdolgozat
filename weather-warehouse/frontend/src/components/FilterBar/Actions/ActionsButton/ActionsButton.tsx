import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown';
import { Box, Tooltip } from "@mui/material";
import React from "react";
import { StyledMenuButton } from '../../../../stlyes/button.style';
import { ActionsMenu } from '../ActionsMenu/ActionsMenu';
import { ActionsButtonProps } from './ActionsButton.type';


export function ActionsButton({ onSaveLocation, onExportLocation, onResetLocation }: ActionsButtonProps) {
  const [anchorElUser, setAnchorElUser] = React.useState<null | HTMLElement>(null);
  const isLoggedIn = !!localStorage.getItem("token");

  function handleOpenUserMenu(event: React.MouseEvent<HTMLElement>) {
    setAnchorElUser(event.currentTarget);
  };

  function handleCloseMenu() {
    setAnchorElUser(null);
  };

  function renderContent() {
    if (!isLoggedIn) {
      return null;
    }

    return (
      <Box sx={{ flexGrow: 0 }} data-testid="actionsButton">
        <Tooltip title="Open actions">
          <StyledMenuButton
            onClick={handleOpenUserMenu}
            endIcon={<KeyboardArrowDownIcon />}
            sx={{ boxShadow: 4 }}
          >
            Actions
          </StyledMenuButton>
        </Tooltip>
        <ActionsMenu
          handleCloseMenu={handleCloseMenu}
          anchorElUser={anchorElUser}
          onSaveLocation={onSaveLocation}
          onExportLocation={onExportLocation}
          onResetLocation={onResetLocation}
        />
      </Box>
    );
  }

  return renderContent();
}
