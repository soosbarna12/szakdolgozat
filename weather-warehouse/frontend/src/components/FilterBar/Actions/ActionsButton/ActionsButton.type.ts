export interface ActionsButtonProps {
  onSaveLocation?: () => void;
}
