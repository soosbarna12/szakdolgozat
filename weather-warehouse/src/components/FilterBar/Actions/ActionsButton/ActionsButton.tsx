import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown';
import { Box, Button, Tooltip } from "@mui/material";
import React from "react";
import { ActionsMenu } from '../ActionsMenu/ActionsMenu';

export function ActionsButton() {
  const [anchorElUser, setAnchorElUser] = React.useState<null | HTMLElement>(null);

  const handleOpenUserMenu = (event: React.MouseEvent<HTMLElement>) => { setAnchorElUser(event.currentTarget); };
  const handleCloseMenu = () => { setAnchorElUser(null); };

  return (
    <Box sx={{ flexGrow: 0 }}>
      <Tooltip title="Open actions">
        <Button
          onClick={handleOpenUserMenu}
          //variant="outlined"
          //color='primary'
          endIcon={<KeyboardArrowDownIcon />}
          sx={{ boxShadow: 4, margin: 1, borderRadius: 20, width: 150, height: 40, fontWeight: 100 }}
        >Actions
        </Button>
      </Tooltip>
      <ActionsMenu
        handleCloseMenu={handleCloseMenu}
        anchorElUser={anchorElUser} />
    </Box>
  );
}