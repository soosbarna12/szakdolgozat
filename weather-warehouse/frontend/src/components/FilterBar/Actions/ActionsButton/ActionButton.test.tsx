import '@testing-library/jest-dom';
import { fireEvent, screen, waitFor } from '@testing-library/react';
import React from 'react';
import { renderWithQueryClient } from '../../../../utils/test/renderWithQueryClient';
import { ActionsButton } from './ActionsButton';


jest.mock('../ActionsMenu/ActionsMenu', () => ({
    ActionsMenu: ({ handleCloseMenu, onSaveLocation, onExportLocation, onResetLocation }: any) => (
        <div data-testid="actionsMenu">
            <button data-testid="saveLocationButton" onClick={onSaveLocation}>Save Location</button>
            <button data-testid="exportLocationButton" onClick={onExportLocation}>Export Location</button>
            <button data-testid="resetLocationButton" onClick={onResetLocation}>Reset Location</button>
            <button data-testid="closeMenuButton" onClick={handleCloseMenu}>Close Menu</button>
        </div>
    ),
}));

jest.mock('../../../../stlyes/button.style', () => ({
    StyledMenuButton: ({ children, endIcon, ...props }: any) => (
        <button {...props} data-testid="styledMenuButton">
            {children}
            {endIcon && <span data-testid="endIcon">{endIcon}</span>}
        </button>
    ),
}));

describe('FilterBar/Actions/ActionsButton', () => {
    const mockOnSaveLocation = jest.fn();
    const mockOnExportLocation = jest.fn();
    const mockOnResetLocation = jest.fn();

    beforeEach(() => {
        jest.clearAllMocks();
        localStorage.setItem('token', 'mockToken');
    });

    afterEach(() => {
        localStorage.removeItem('token');
    });

    it('matches ActionsButton component snapshot', () => {
        const component = renderWithQueryClient(
            <ActionsButton
                onSaveLocation={mockOnSaveLocation}
                onExportLocation={mockOnExportLocation}
                onResetLocation={mockOnResetLocation}
            />
        );

        expect(component).toMatchSnapshot();
    });

    it('does not render anything when the user is not logged in', () => {
        localStorage.removeItem('token');

        renderWithQueryClient(
            <ActionsButton
                onSaveLocation={mockOnSaveLocation}
                onExportLocation={mockOnExportLocation}
                onResetLocation={mockOnResetLocation}
            />
        );

        expect(screen.queryByTestId('styledMenuButton')).not.toBeInTheDocument();
    });

    it('renders the menu button and opens the menu on click', () => {
        renderWithQueryClient(
            <ActionsButton
                onSaveLocation={mockOnSaveLocation}
                onExportLocation={mockOnExportLocation}
                onResetLocation={mockOnResetLocation}
            />
        );

        const menuButton = screen.getByTestId('styledMenuButton');
        fireEvent.click(menuButton);

        expect(screen.getByTestId('actionsMenu')).toBeInTheDocument();
    });

    it('calls onSaveLocation when the save location button is clicked', async () => {
        renderWithQueryClient(
            <ActionsButton
                onSaveLocation={mockOnSaveLocation}
                onExportLocation={mockOnExportLocation}
                onResetLocation={mockOnResetLocation}
            />
        );

        const menuButton = screen.getByTestId('styledMenuButton');
        fireEvent.click(menuButton);

        const saveButton = screen.getByTestId('saveLocationButton');
        fireEvent.click(saveButton);

        await waitFor(() => {
            expect(mockOnSaveLocation).toHaveBeenCalled();
        });
    });

    it('calls onExportLocation when the export location button is clicked', async () => {
        renderWithQueryClient(
            <ActionsButton
                onSaveLocation={mockOnSaveLocation}
                onExportLocation={mockOnExportLocation}
                onResetLocation={mockOnResetLocation}
            />
        );

        const menuButton = screen.getByTestId('styledMenuButton');
        fireEvent.click(menuButton);

        const exportButton = screen.getByTestId('exportLocationButton');
        fireEvent.click(exportButton);

        await waitFor(() => {
            expect(mockOnExportLocation).toHaveBeenCalled();
        });
    });

    it('calls onResetLocation when the reset location button is clicked', async () => {
        renderWithQueryClient(
            <ActionsButton
                onSaveLocation={mockOnSaveLocation}
                onExportLocation={mockOnExportLocation}
                onResetLocation={mockOnResetLocation}
            />
        );

        const menuButton = screen.getByTestId('styledMenuButton');
        fireEvent.click(menuButton);

        const resetButton = screen.getByTestId('resetLocationButton');
        fireEvent.click(resetButton);

        await waitFor(() => {
            expect(mockOnResetLocation).toHaveBeenCalled();
        });
    });

});