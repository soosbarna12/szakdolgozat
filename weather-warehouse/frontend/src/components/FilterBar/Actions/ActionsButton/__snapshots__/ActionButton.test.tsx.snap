// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBar/Actions/ActionsButton matches ActionsButton component snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-2uchni"
        data-testid="actionsButton"
      >
        <button
          aria-label="Open actions"
          class=""
          data-mui-internal-clone-element="true"
          data-testid="styledMenuButton"
          sx="[object Object]"
        >
          Actions
          <span
            data-testid="endIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="KeyboardArrowDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
              />
            </svg>
          </span>
        </button>
        <div
          data-testid="actionsMenu"
        >
          <button
            data-testid="saveLocationButton"
          >
            Save Location
          </button>
          <button
            data-testid="exportLocationButton"
          >
            Export Location
          </button>
          <button
            data-testid="resetLocationButton"
          >
            Reset Location
          </button>
          <button
            data-testid="closeMenuButton"
          >
            Close Menu
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-2uchni"
      data-testid="actionsButton"
    >
      <button
        aria-label="Open actions"
        class=""
        data-mui-internal-clone-element="true"
        data-testid="styledMenuButton"
        sx="[object Object]"
      >
        Actions
        <span
          data-testid="endIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="KeyboardArrowDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
            />
          </svg>
        </span>
      </button>
      <div
        data-testid="actionsMenu"
      >
        <button
          data-testid="saveLocationButton"
        >
          Save Location
        </button>
        <button
          data-testid="exportLocationButton"
        >
          Export Location
        </button>
        <button
          data-testid="resetLocationButton"
        >
          Reset Location
        </button>
        <button
          data-testid="closeMenuButton"
        >
          Close Menu
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
